Add RFC 9253 compatibility

niccokunzmann opened the pull request:
## Closes issue

Replace `ISSUE_NUMBER` with the issue number that your pull request fixes. Then GitHub will link and automatically close the related issue.

- [x] Closes #658

## Description

This intends to solve the issue

## Checklist

- [x] add parameter types "# parameter types (luckily there is no name overlap)"
- [x] add to new()
- [x] figure out why the type hints do not work for the properties like `Event.conferences` and `Event.links`
- [x] I've added a change log entry to `CHANGES.rst`.
- [x] I've added or updated tests if applicable.
- [x] I've run and ensured all tests pass locally by following [Running Tests](https://icalendar.readthedocs.io/en/latest/install.html#running-tests).
- [x] I've added or edited documentation, both as docstrings to be rendered in the API documentation and narrative documentation, as necessary.
- [x] I've added myself to `docs/credits.rst` as a contributor in this pull request or have done so previously.

Still TODO:

- [x] https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.html#icalendar.vUri.GAP timedelta should be clickable.
- [x] timezone link is sometimes not available - https://github.com/collective/icalendar/actions/runs/19035348837/job/54358219688#step:6:299: check validity and then auto-accept it (ignore)
  ```
  (reference/api/icalendar.timezone.windows_to_olson: line    8) broken    https://www.unicode.org/cldr/cldr-aux/charts/29/supplemental/zone_tzid.html - ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
  ```

## Documentation Review guidance

### attr.py

- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.attr.html#icalendar.attr.multi_string_property

### icalendar/__init__.py

- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.html#icalendar.vText
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.html#icalendar.vUid
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.html#icalendar.vUri
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.html#icalendar.vXmlReference

### enums.py

- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.enums.html#icalendar.enums.RELTYPE

### param.py

- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.param.html#icalendar.param.FMTTYPE  # Doesn't appear. Should it?
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.param.html#icalendar.param.LABEL  # Doesn't appear. Should it?
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.param.html#icalendar.param.LINKREL  # Doesn't appear. Should it?
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.param.html#icalendar.param.RELTYPE  # Doesn't appear. Should it?
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.param.html#icalendar.param.VALUE  # Doesn't appear. Should it?

### parser.py

- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.parser.html#icalendar.parser.Parameters

### prop/__init__.py

- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.prop.html#icalendar.prop.vUid
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.prop.html#icalendar.prop.vUri
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.prop.html#icalendar.prop.vXmlReference

### Components

- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.html#icalendar.cal.Component.links
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.html#icalendar.cal.Component.concepts
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.html#icalendar.cal.Component.related_to

### `new()`

- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.alarm.html#icalendar.cal.alarm.Alarm.new
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.availability.html#icalendar.cal.availability.Availability.new
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.available.html#icalendar.cal.available.Available.new
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.calendar.html#icalendar.cal.calendar.Calendar.new
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.event.html#icalendar.cal.event.Event.new
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.free_busy.html#icalendar.cal.free_busy.FreeBusy.new
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.journal.html#icalendar.cal.journal.Journal.new
- https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.cal.todo.html#icalendar.cal.todo.Todo.new

### Change log

- https://icalendar--941.org.readthedocs.build/en/941/reference/changelog.html#a2-unreleased


## Additional information

https://www.rfc-editor.org/rfc/rfc9253.html


## TODO

- after merge
  - [ ] document how to properly link to objects, classes and functions https://github.com/collective/icalendar/issues/991#issuecomment-3536171001
  - [ ] document how to use the new functionality https://github.com/collective/icalendar/issues/997

stevepiercy commented:
Sorry, $WORK jumped all over me yesterday. I'll do review this today.

stevepiercy commented on CHANGES.rst:
```diff
@@ -32,6 +32,16 @@ Breaking changes:
+  - Add ``LABEL``, ``GAP``, and ``VALUE`` parameter to property values affected by :rfc:`9253`.
+
+- Allow setting default values in :class:`icalendar.parser.Parameters`.
+- Allow ``None`` as an argument to :class:`icalendar.parser.Parameters`.
```
```suggestion
- Allow setting default values in :py:class:`icalendar.parser.Parameters`.
- Allow ``None`` as an argument to :py:class:`icalendar.parser.Parameters`.
```

stevepiercy commented on docs/conf.py:
```diff
@@ -141,6 +142,7 @@
     r"https://pypi.org/manage/project/icalendar/collaboration/",
     # Ignore specific anchors
     r"https://github.com/collective/icalendar/blob/main/README.rst#related-projects",
+    r"https://www.unicode.org/cldr/cldr-aux/charts/29/supplemental/zone_tzid.html",
```
Why do you want to ignore this link from the checker? It also lacks an anchor, which the comment above implies it should have.

```suggestion
```

niccokunzmann replied:
It breaks from time to time on github CI. I can move it up.

stevepiercy commented on docs/conf.py:
```diff
@@ -21,6 +21,7 @@
     "sphinx.ext.intersphinx",
     "sphinx.ext.autosectionlabel",
     "sphinx.ext.napoleon",
+    "sphinx_autodoc_typehints",
```
When adding a new Sphinx extension, it should be documented where appropriate. In this case, update https://icalendar.readthedocs.io/en/latest/contribute/development.html#apidoc with a new typehint docstring example.

niccokunzmann replied:
Oh. That is just there so that the type aliases get resolved.
The docstring there is typed already. What would you like to see? At the moment, I do not know how to follow your request.

niccokunzmann replied:
No, it is duplicated code to type the arguments inside the string. This is the solution to resolving TypeAlias variables to their actual value:

https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.html#icalendar.Calendar.new
<img width="600" height="53" alt="grafik" src="https://github.com/user-attachments/assets/790c1f15-a1a6-4cea-89bf-97fd8f85b6c6" />

It seems they are automatically embedded and resolved.

stevepiercy commented on src/icalendar/attr.py:
```diff
@@ -644,7 +652,11 @@ def _del_categories(component: Component) -> None:
+
+.. seealso::
+
+    :attr:`Component.concepts`
```
This doesn't render, and I don't know why. See https://icalendar.readthedocs.io/en/latest/reference/api/icalendar.html#icalendar.Availability.categories and https://icalendar.readthedocs.io/en/latest/reference/api/icalendar.cal.component.html. Is this even a thing?

```suggestion
.. seealso::

    :py:attr:`Component.concepts`
```

stevepiercy commented on src/icalendar/attr.py:
```diff
@@ -1977,13 +1989,452 @@ def _del_conferences(self: Component):
+        - LANGUAGE
+            This parameter maps to the "hreflang" attribute defined in Section 3.4.1
+            of :rfc:`8288`. See :rfc:`5646`. Example: ``en``, ``de-ch``.
+        - LINKREL
```
```suggestion
        LINKREL
```

stevepiercy commented on src/icalendar/attr.py:
```diff
@@ -1977,13 +1989,452 @@ def _del_conferences(self: Component):
+             https://example.com/caldav/user/jb/cal/
+             19960401-080045-4000F192713.ics
+
+    See also :class:`icalendar.enum.RELTYPE`.
```
```suggestion
    See also :py:class:`icalendar.enum.RELTYPE`.
```

stevepiercy commented on src/icalendar/attr.py:
```diff
@@ -1977,13 +1989,452 @@ def _del_conferences(self: Component):
+
+    .. seealso::
+
+        :attr:`Compnent.categories`
```
```suggestion
        :py:attr:`Component.categories`
```

stevepiercy commented on src/icalendar/cal/component.py:
```diff
@@ -71,14 +82,19 @@ def get_component_class(cls, name: str) -> type[Component]:
     @staticmethod
-    def _infer_value_type(value: date | datetime | timedelta | time | tuple | list) -> str | None:
+    def _infer_value_type(
+        value: date | datetime | timedelta | time | tuple | list,
```
Why a trailing comma? Looks like an overly eager formatter pooped a comma in your code.

Please resist reformatting code that you otherwise wouldn't touch. It bloats PRs. A separate PR that only reformats the entire code base is easier to manage.

stevepiercy commented on src/icalendar/cal/free_busy.py:
```diff
@@ -105,9 +112,16 @@ def new(
 
         Arguments:
             comments: The :attr:`Component.comments` of the component.
```
Moar `:py` prefixes needed.

stevepiercy commented on src/icalendar/cal/todo.py:
```diff
@@ -320,13 +327,18 @@ def new(
             classification: The :attr:`classification` of the todo.
             color: The :attr:`color` of the todo.
             comments: The :attr:`Component.comments` of the todo.
+            concepts: The :attr:`Component.concepts` of the todo.
```
Moar `:py` prefixes needed.

stevepiercy commented on src/icalendar/enums.py:
```diff
@@ -199,11 +208,77 @@ class RELTYPE(StrEnum):
+        * RELTYPE=CONCEPT
+            This establishes a reference from the current component to
+            components with a CONCEPT property that matches the value
+            given in the associated RELATED-TO property.
```
Convert the above two unordered lists to definition lists by removing `* ` from each term.

stevepiercy reviewed (changes requested):
All Python objects need to have markup that has the prefix of `:py` to render as a link instead of an inline literal. Example:

```rst
            concepts: The :py:attr:`Component.concepts` of the availability.
```

I started doing each one of these, but realized that would be inefficient, so I switched to doing it only once per method, class, or other Python object. I wouldn't change them everywhere, but perhaps as a separate PR where formatting code is its only purpose.

Before you actually make these `:py` changes, I'd suggest that you test it with one change only to ensure that it actually generates the rendered link as I expect it will. I could be wrong, and it'd be annoying to go through all that work for nothing.

niccokunzmann commented on CHANGES.rst:
```diff
@@ -32,6 +32,16 @@ Breaking changes:
+  
+  - Add new property types :class:`vUid` and :class:`vXmlReference`.
+  - Add properties to all components: :attr:`Component.concepts`, :attr:`Component.links`, :attr:`Component.refids`, :attr:`Component.related_to`.
+  - Add new values to :class:`icalendar.enum.RELTYPE`.
```
Hm. Adding `:py` does not change the issue. This issue addresses this: https://github.com/collective/icalendar/issues/994

niccokunzmann commented:
@stevepiercy, there are two unresolved comments and I opened issue #994. What are your thoughts on how to progress?

stevepiercy commented on CHANGES.rst:
```diff
@@ -32,6 +32,16 @@ Breaking changes:
+  - Add new property types :py:class:`vUid` and :py:class:`vXmlReference`.
+  - Add properties to all components: :attr:`Component.concepts`, :py:attr:`Component.links`, :py:attr:`Component.refids`, :py:attr:`Component.related_to`.
+  - Add new values to :py:class:`icalendar.enum.RELTYPE`.
+  - Add ``LABEL``, ``GAP``, and ``VALUE`` parameter to property values affected by :rfc:`9253`.
```
```suggestion
  - Add ``LABEL``, ``GAP``, and ``VALUE`` parameters to property values affected by :rfc:`9253`.
```

stevepiercy commented on CHANGES.rst:
```diff
@@ -32,6 +32,16 @@ Breaking changes:
 
+- Add compatibility to :rfc:`9253`:
+  
+  - Add new property types :py:class:`vUid` and :py:class:`vXmlReference`.
```
It looks like my suggestion to add `:py` prefix was wrong. There's no links in https://icalendar--941.org.readthedocs.build/en/941/reference/changelog.html#a2-unreleased.

I think you found that we need to use this syntax for icalendar objects instead, and we need to use `:py` only for Python standard library objects?

```suggestion
  - Add new property types :class:`icalender.vUid` and :class:`icalender.vXmlReference`.
```

niccokunzmann replied:
> I think it would be good to have an example of each way to link to the various icalendar and standard library objects.

Yeah... Documented! :heart:

stevepiercy replied:
These still don't link to the icalendar classes or methods. Unresolving.

stevepiercy commented on src/icalendar/parser.py:
```diff
@@ -206,10 +206,57 @@ def fdel(self: Parameters):
+
+        Set a default:
+
+       .. code-block:: pycon
```
Funky format. See third example in https://icalendar--941.org.readthedocs.build/en/941/reference/api/icalendar.parser.html#icalendar.parser.Parameters

```suggestion
        .. code-block:: pycon
```

stevepiercy commented:
@niccokunzmann I resolved one of the two comments, but see https://github.com/collective/icalendar/pull/941#discussion_r2528959130.

I also added another review.

I didn't do a thorough job of checking the additions in the API documentation. I'm not sure where these things should appear, and I got lost.

SashankBhamidi commented on src/icalendar/prop/__init__.py:
```diff
@@ -1808,6 +1790,100 @@ def from_ical(cls, ical):
+    def __init__(self, xml_reference: str):
+        """Create a new XML reference."""
+        super().__init__()
+        self.params.setdefault("VALUE", "XML-REFERENCE")
```
```suggestion
    def __init__(self, xml_reference: str):
        """Create a new XML reference."""
        super().__init__(xml_reference)
        self.params.setdefault("VALUE", "XML-REFERENCE")
```

Same issue as `vUid` - the `xml_reference` parameter should be passed to parent.

SashankBhamidi commented on src/icalendar/prop/__init__.py:
```diff
@@ -1808,6 +1790,100 @@ def from_ical(cls, ical):
+    @property
+    def xml_reference(self) -> str:
+        """The uid of this property."""
+        return self.uri
```
```suggestion
    @property
    def xml_reference(self) -> str:
        """The XML reference URI of this property."""
        return self.uri
```

Docstring says "The uid of this property" but should say "The XML reference URI of this property".

SashankBhamidi commented on src/icalendar/param.py:
```diff
@@ -476,6 +485,243 @@ def _default_return_parent() -> enums.RELTYPE:
+    """GAP
+
+    Purpose:
+        GAP pecifies the length of the gap, positive or negative,
```
```suggestion
        GAP specifies the length of the gap, positive or negative,
```

Missing 's' in "specifies".

SashankBhamidi commented on src/icalendar/prop/__init__.py:
```diff
@@ -1808,6 +1790,100 @@ def from_ical(cls, ical):
+
+    @property
+    def xml_reference(self) -> str:
+        """The uid of this property."""
```
```suggestion
        """The XML reference URI of this property."""
```

Says "uid" but this is the xml_reference property.

niccokunzmann commented on CHANGES.rst:
True. That is a good idea. There could be a section that explains how to use the compatibility with the RFC.

stevepiercy replied:
Most projects keep change logs simple, due to automation, such as generation from commit messages and other processes.

An upgrade guide is a damned good idea. Here are a couple of links to good examples:

- https://6.docs.plone.org/backend/upgrading/index.html
- https://6.docs.plone.org/volto/upgrade-guide/index.html
- https://6.docs.plone.org/plone.restapi/docs/source/upgrade-guide.html
- https://6.docs.plone.org/upgrade/index.html

I created an issue to track it https://github.com/collective/icalendar/issues/997, so that it doesn't hold up this PR. Please add specific items that ought to be added to the upgrade guide when work starts on it. @niccokunzmann @SashankBhamidi

niccokunzmann commented on src/icalendar/param.py:
```diff
@@ -476,6 +485,243 @@ def _default_return_parent() -> enums.RELTYPE:
+
+def _del_GAP(prop):  # noqa: N802
+    """Delete the GAP parameter."""
+    prop.params.pop("GAP", None)
```
This is addressed by be1b694
Since vDuration comes from parsing and setters, I think, type checking is ok here.

niccokunzmann commented:
@stevepiercy @SashankBhamidi , thank you for the reviews. I think, I addressed all the feedback. I added a todo for two issues at the top. I would re-request the review so you can feedback or accept.

stevepiercy commented on CHANGES.rst:
```diff
@@ -34,7 +34,7 @@ New features:
 - Add compatibility to :rfc:`9253`:
   
-  - Add new property types :py:class:`vUid` and :py:class:`vXmlReference`.
+  - Add new property types :class:`icalender.vUid` and :class:`icalender.vXmlReference`.
```
Fix typos. Maybe that's why they still don't link?

```suggestion
  - Add new property types :class:`icalendar.vUid` and :class:`icalendar.vXmlReference`.
```

stevepiercy commented on CHANGES.rst:
```diff
@@ -32,6 +32,16 @@ Breaking changes:
+- Add compatibility to :rfc:`9253`:
+  
+  - Add new property types :class:`icalender.vUid` and :class:`icalender.vXmlReference`.
+  - Add properties to all components: :attr:`Component.concepts`, :py:attr:`Component.links`, :py:attr:`Component.refids`, :py:attr:`Component.related_to`.
```
All these should have consistent markup and link to the relevant attr. Is there an example of valid reST markup for `attr`?

niccokunzmann commented on docs/conf.py:
```diff
@@ -21,6 +21,7 @@
     "sphinx.ext.intersphinx",
     "sphinx.ext.autosectionlabel",
     "sphinx.ext.napoleon",
+    "sphinx_autodoc_typehints",
```
Hm. I am not sure if it had an effect but I like the way it looks now...

stevepiercy replied:
If you're happy, I'm happy. 😃

stevepiercy reviewed (changes requested):
Just a couple minor things to update, and I think we're good to merge.

I updated the list of docs in the description to verify docstrings render as expected. I wasn't sure about the ones under param.py, but figured it would be good to ask.

Thank you!

stevepiercy commented on src/icalendar/param.py:
```diff
@@ -476,6 +485,245 @@ def _default_return_parent() -> enums.RELTYPE:
+    :rfc:`9253` makes use of this for the LINK property:
+    This parameter maps to the "title"
+    attribute defined in Section 3.4.1 of :rfc:`8288`:
+    LABEL is used to label the destination
```
```suggestion
    :rfc:`9253` makes use of this for the LINK property.
    This parameter maps to the "title"
    attribute defined in Section 3.4.1 of :rfc:`8288`.
    LABEL is used to label the destination
```
